src/Microsoft.VisualStudio.Threading.Analyzers.CodeFixes/VSTHRD201CancelAfterSwitchToMainThreadCodeFix.cs

```diff
@@ -0,0 +1,100 @@
+﻿/********************************************************
+*                                                        *
+*   © Copyright (C) Microsoft. All rights reserved.      *
+*                                                        *
+*********************************************************/
+
+namespace Microsoft.VisualStudio.Threading.Analyzers
+{
+    using System;
+    using System.Collections.Immutable;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.CodeAnalysis;
+    using Microsoft.CodeAnalysis.CodeActions;
+    using Microsoft.CodeAnalysis.CodeFixes;
+    using Microsoft.CodeAnalysis.CSharp;
+    using Microsoft.CodeAnalysis.CSharp.Syntax;
+    using Microsoft.CodeAnalysis.Formatting;
+
+    [ExportCodeFixProvider(LanguageNames.CSharp)]
+    public class VSTHRD201CancelAfterSwitchToMainThreadCodeFix : CodeFixProvider
+    {
+        private static readonly ImmutableArray<string> ReusableFixableDiagnosticIds = ImmutableArray.Create(
+            VSTHRD201CancelAfterSwitchToMainThreadAnalyzer.Id);
+
+        /// <inheritdoc />
+        public override ImmutableArray<string> FixableDiagnosticIds => ReusableFixableDiagnosticIds;
+
+        /// <inheritdoc />
+        public override FixAllProvider GetFixAllProvider() => WellKnownFixAllProviders.BatchFixer;
+
+        /// <inheritdoc />
+        public override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var diagnostic = context.Diagnostics.First();
+
+            // Our fix only works if we're within a block or if statement (no simple lambdas),
+            // so check applicability before offering.
+            var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            var statementSyntax = root.FindNode(diagnostic.Location.SourceSpan).FirstAncestorOrSelf<StatementSyntax>();
+            if (statementSyntax?.Parent is BlockSyntax || statementSyntax?.Parent is IfStatementSyntax)
+            {
+                context.RegisterCodeFix(
+                    CodeAction.Create(
+                        Strings.VSTHRD201_CodeFix_Title,
+                        ct => this.AddThrowOnCanceledAsync(context, diagnostic, ct),
+                        equivalenceKey: nameof(CancellationToken.ThrowIfCancellationRequested)),
+                    diagnostic);
+            }
+        }
+
+        private static ExpressionSyntax? GetCancellationTokenInInvocation(InvocationExpressionSyntax invocation, SemanticModel semanticModel, INamedTypeSymbol cancellationTokenTypeSymbol)
+        {
+            // Consider that named arguments allow for alternative ordering.
```

Comment:
💡 The operation will allow you to match a known parameter. Consider adding the actual argument syntax index to the property bag for the diagnostic so you don't have to try and match it again here using only the syntax.

Reply:
Done.